Social Login via API

timothyej opened the pull request:
Fixes #112.

This PR adds support for social login from SPAs with for example the Angular SDK.

### Includes

* Support for login using `providerData` instead of username/password with the login API
* A new API route, `/oauth/providers`, that returns a list of all OAuth providers and their configurations
* Logouts with the `Accept` header set to `application/json` will respond with 200 instead of a redirect

### Notes

* Is the route `/oauth/providers` the best route for this purpose?

### How to verify

1. Setup a server with this branch
2. Verify the login, logout, etc. works with the AngularJS SDK: <https://github.com/stormpath/stormpath-sdk-angularjs/pull/48>
3. Run tests

rdegges commented:
Looking good so far =)

robertjd commented:
I've reviewed the code changes and I like it, I don't have any comments on the code.  But please add the following to this PR:

* JSDoc comments for new functions

* Tests for these cases:

 * That the `/oauth/providers` endpoint is serving the expected object, if the application is mapped to a directory that has provider data

 * That the `/oauth/providers` endpoint returns an empty object, `{}`, when there are no account stores with provider data on the application

timothyej commented:
Thanks @robertjd, I will fix the docs and tests then this should be ready for final review :)

timothyej commented:
@rdegges @robertjd @typerandom This is ready for final review! I've added some jsdocs and some tests and fixed the conflicts. I don't know why the Travis build fails though?

typerandom commented on lib/controllers/login.js:
```diff
@@ -6,7 +6,8 @@ var uuid = require('uuid');
 var forms = require('../forms');
 var helpers = require('../helpers');
-var oauth = require('../oauth');
+var oauth = require('../oauth')
```
Missing `;`.

timothyej replied:
Good catch :+1:

typerandom commented on lib/controllers/login.js:
```diff
@@ -50,6 +51,19 @@ module.exports = function(req, res) {
   res.locals.status = req.query.status;
 
   if (req.method === 'POST' && accepts === 'json') {
+    if (req.body.providerData) {
```
How about adding `var providerData = req.body.providerData;` then `if (providerData) {`, etc, and using that instead of repeating it?

timothyej replied:
Much better :)

typerandom commented on lib/controllers/logout.js:
```diff
@@ -14,11 +14,17 @@ var middleware = require('../middleware');
  */
 module.exports = function(req, res) {
   var config = req.app.get('stormpathConfig');
+  var accepts = req.accepts([ 'json' ]);
```
That array style looks very inconsistent. Haven't seen it before.

timothyej replied:
`login.js` uses this style, but I will change them both to not use spaces. I also decided to add `html` to the list to be consistent and be clear that this route supports both html and json.

typerandom commented on test/controllers/test-logout.js:
```diff
@@ -52,4 +52,26 @@ describe('logout', function() {
       .end(done);
   });
+
+  describe('when Accept header is set to text/html', function() {
```
Looks inconsistent to use mixed case for describe/it statements.

timothyej replied:
Why, just because it starts with lowercase doesn't mean all letters have to be lowercase? The accept header is written with a capital `A` so it only makes sense in my opinion.

typerandom replied:
You're right. My bad. I was just used to lowercase being used as the general style, but I can see now that it's being used differently.

typerandom commented on test/controllers/test-oauth-providers.js:
```diff
@@ -0,0 +1,151 @@
+    before(performRequest);
+
+    it('should return an object', function() {
+      assert.equal(typeof response === 'object', true);
```
Clearer to use `assert.typeOf`.

timothyej replied:
Nice, didn't know about that one :+1:

typerandom commented on test/controllers/test-oauth-providers.js:
```diff
@@ -0,0 +1,151 @@
+    });
+
+    it('should have "facebook.enabled" set to false', function() {
+      assert.equal(response.hasOwnProperty('facebook'), true);
```
Clearer to use `assert.isTrue`.

timothyej replied:
Btw, which assertion library are we using? Because I can't find these functions here: <https://nodejs.org/api/assert.html> :(

typerandom replied:
Chai.

http://chaijs.com/api/assert/

timothyej replied:
Oh, that wasn't very clear 😂

typerandom replied:
Actually, when looking at this now it seems like in `stormpath-express` Node.js built in assertion library isn't being used. We only use Chai in the SDK. Which is very inconsistent.

typerandom commented on test/controllers/test-oauth-providers.js:
```diff
@@ -0,0 +1,151 @@
+
+    it('should have "facebook.enabled" set to false', function() {
+      assert.equal(response.hasOwnProperty('facebook'), true);
+      assert.equal(response.facebook.enabled, false);
```
Clearer to use `assert.isFalse`.

timothyej commented:
Thanks for the review, @typerandom! I've addressed your comments in my latest commit.

robertjd commented:
Closing in favor of #171
